src/main/java/com/google/gwt/dom/client/DOMImpl.java

```diff
@@ -19,15 +19,108 @@
 
 import com.google.common.base.MoreObjects;
 import com.google.gwt.core.client.JsArray;
+import com.google.gwt.event.dom.client.BlurEvent;
+import com.google.gwt.event.dom.client.ChangeEvent;
 import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.FocusEvent;
+import com.google.gwt.event.dom.client.DomEvent.Type;
+import com.google.gwt.user.client.DOM;
+import com.google.gwt.user.client.Event;
+import com.google.gwt.user.client.EventListener;
 
 public class DOMImpl {
 	
 	static final DOMImpl impl = new DOMImpl();
 	
-	public void buttonClick(ButtonElement button) {
-		button.fireEvent(ClickEvent.getType());
-	}
+	// Element In Focus (some browsers actually have working implementation document.activeElement)
+	private Element activeElement;
+	
+    private void switchFocusTo(Element element) {
+        if (activeElement == element) {
+            return;
+        } else {
+            if (activeElement != null && isAttached(activeElement)) {
+                if (activeElement instanceof InputElement) {
+                    fireValueChangeIfAny((InputElement)activeElement);
+                }
+                activeElement.fireEvent(BlurEvent.getType());
+            }
+            element.fireEvent(FocusEvent.getType());
+            activeElement = element;
+        }
+    }
+
+    private boolean isAttached(Element elem) {
+        return DOM.getEventListener(elem) != null;
+    }
+
+    private void fireValueChangeIfAny(InputElement element) {
```

Comment:
I see you have a very solid point. I wish I new it from the start.

I'm very hesitant to create an maintain a new project ' gwt-mock-helpers'.  Can it be a part of this project as a different sub-module or a new Class?
 The problem is that in many case it will rely anyway on mock implementation and changers to 'gwt-mock-2.5.1'  will be required at the same time.

-- 

I will try to start all over in a new PR.